WooFoundation/WooFoundation/Extensions/Date+StartAndEnd.swift

```diff
@@ -96,6 +96,64 @@ public extension Date {
         return calendar.date(byAdding: components, to: startOfMonth)
     }
 
+    // MARK: Quarter
+
+    /// Returns self's start of quarter in the given time zone.
+    func startOfQuarter(timezone: TimeZone, calendar: Calendar) -> Date? {
+        guard let startOfMonth = startOfMonth(timezone: timezone) else {
+            return nil
+        }
+
+        var components = calendar.dateComponents([.month, .year], from: startOfMonth)
+
+        let quarterFirstMonth: Int
+        switch components.month {
+        case 1, 2, 3:
+            quarterFirstMonth = 1
+        case 4, 5, 6:
+            quarterFirstMonth = 4
+        case 7, 8, 9:
+            quarterFirstMonth = 7
+        case 10, 11, 12:
+            quarterFirstMonth = 10
+        default:
+            return nil
+        }
+        components.month = quarterFirstMonth
+        return calendar.date(from: components)
+    }
+
+    /// Returns self's end of quarter in the given time zone.
+    func endOfQuarter(timezone: TimeZone, calendar: Calendar) -> Date? {
+        guard let startOfMonth = startOfMonth(timezone: timezone) else {
+            return nil
+        }
+
+        var components = calendar.dateComponents([.month, .day, .year], from: startOfMonth)
+
+        switch components.month {
+        case 1, 2, 3:
+            components.month = 3
+        case 4, 5, 6:
+            components.month = 6
+        case 7, 8, 9:
+            components.month = 9
+        case 10, 11, 12:
+            components.month = 12
+        default:
+            return nil
+        }
+
+        guard let quarterLastMonth = calendar.date(from: components) else {
+            return nil
+        }
+
+        var oneMonthUnit = DateComponents()
+        oneMonthUnit.month = 1
+        oneMonthUnit.second = -1
+        return calendar.date(byAdding: oneMonthUnit, to: quarterLastMonth)
+    }
+
```

Comment:
I wanted to avoid this manual implementation around quarters, but the calendar component for quarters simply doesn't work. I tried to verify possible solutions, and even Apple documentation states that the [quarter unit is `largely unimplemented`](https://developer.apple.com/documentation/foundation/calendar/component/quarter).

Let me know if you folks knows better ways for us to operate the quarter data.